### DISCUSSION TITLE
Use `PortableGL` for software rendering

### DISCUSSION THREAD

**Discussion by raysan5** [raysan5/raylib#2036] [Ideas]
[`PortableGL`](https://github.com/rswinkle/PortableGL) is a direct replacement for OpenGL libraries to allow OpenGL 3.3 rendering in software.

Some concerns:  

 - raylib supports OpenGL ES 2.0, could PortableGL support it? (it's quite similar to OpenGL 3.3 but it has some differences)
 - As listed in the mentioned issue, the main problem in raylib for a software rederer is the non-accelerated Window required, SDL2 can create it but GLFW used in raylib can't. Also, all the other platforms are in the same situation... it would be up to the user to manage it.

Related issue: https://github.com/raysan5/raylib/issues/1370

**Comment by rswinkle** [comment]
>PortableGL is a direct replacement for OpenGL libraries to allow OpenGL 3.3 rendering in software. 

Not quite.  The major difference is how it handles shaders.  Writing a GLSL parser/compiler etc. was way beyond the scope of the project so shaders are just C/C++ functions.  This means all the inputs and outputs to shaders are parameters.  Also textures are accessed by handle, (there are no Texture Units).

Of course there are also many minor functions and features that are lacking but the big one that will always keep it from 3.3 is the Geometry Shader.  Like proper GLSL shader support that's too much for me and for software rendering performance imo.

I do remember that 3.x and ES 2.0 were relatively similar with the latter lacking quite as extensive format support.  Beyond that I'm not familiar.  Is there a good website or document that contrasts them directly?

Yes, the second bullet point is the reason I use SDL2 for my demos; it's the easiest way to get the pixels to the screen.  It doesn't depend on SDL2 but it does require some way to blit the framebuffer to the screen unless you're doing offline rendering which I doubt many use raylib for.

**Comment by KorbisInc** [comment]


**Reply by wwderw** [reply]
I agree having software rendering would be great.  I think for quite a few situations, the hit with software rendering versus using acceleration isn't quite as bad as one thinks.  Plus I too use VMs extensively and this would help with that as well.  I hope this is able to added into Raylib.

**Comment by raysan5** [comment]
As far as I understand it, no need to do anything on raylib side, only link with `PortableGL`, right? Anyone tested it?

**Reply by rswinkle** [reply]
I think the original issue you mentioned a couple years ago still applies.  It can't really be integrated directly because GLFW doesn't support writing directly to the window.

As far as using it directly, you can definitely do that with some minor porting hurdles.  I also recently added a small [built in](https://github.com/rswinkle/PortableGL/blob/master/src/pgl_std_shaders.h) shader [library](https://github.com/rswinkle/PortableGL/blob/master/src/pgl_std_shaders.c) covering some common use cases.

I can answer any questions here or on my own repo if you run into any bugs or critical missing features.

**Reply by rswinkle** [reply]
And ... I just realized I can answer your original question a little more thoroughly now.  While I still describe PortableGL as "3.x ish" it's really more ambiguous than that.  I started with the goal of 3.3 minus some impractical features like the geometry shader but I have added several 4.x functions (like DSA functions) and more recently some OpenGL ES 2 things.  If it's relatively easy add and doesn't affect performance, I'm usually willing to add it.  I've actually been working on and comparing the performance of identical demo programs in GLES 2 and PGL on a RPi 3B+.  It was very easy to port the former to the latter, just write the shaders and deal with the fact PGL only accepts RGBA textures (on my TODO list)

So long story short, yes it can do OpenGL ES 2.

**Comment by TSnake41** [comment]
Maybe Mesa3D could be useful there, although it is quite big.
